internal/v1/handler_compose_image.go

```diff
@@ -467,6 +467,97 @@ func (h *Handlers) buildCustomRepositories(ctx echo.Context, custRepos []CustomR
+	}
+	for _, id := range *template.RepositoryUuids {
+		// Check if repo is Red Hat
+		rhRepoMap, err := h.server.csClient.GetRepositories(ctx.Request().Context(), []string{}, []string{id}, false)
+		if err != nil {
+			return nil, nil, fmt.Errorf("Unable to retrieve Red Hat repositories: %v", err)
+		}
+		// Separate Red Hat repos from custom
+		if len(rhRepoMap) > 0 {
+			continue
+		} else {
+			customRepoIDs = append(customRepoIDs, id)
+		}
+	}
+
+	var repositories []composer.Repository
+	var customRepositories []composer.CustomRepository
+
+	// No custom repos in template, only Red Hat repos
+	if len(customRepoIDs) == 0 {
+		return repositories, customRepositories, nil
+	}
+
+	repoMap, err := h.server.csClient.GetRepositories(ctx.Request().Context(), []string{}, customRepoIDs, true)
+	if err != nil {
+		return nil, nil, fmt.Errorf("Unable to retrieve external repositories: %v", err)
+	}
+
+	// We should never hit this condition, but checking just in case
+	if template.Snapshots == nil {
+		return nil, nil, fmt.Errorf("Template %v has no snapshots", templateID)
+	}
+	for _, snap := range *template.Snapshots {
+		if snap.RepositoryUuid == nil {
+			return repositories, customRepositories, fmt.Errorf("No repository UUID is associated with this snapshot")
+		}
+		if slices.Contains(customRepoIDs, *snap.RepositoryUuid) {
+			repo, ok := repoMap[*snap.RepositoryUuid]
+			if !ok {
+				return repositories, customRepositories, fmt.Errorf("Returned snapshot %v unexpected repository id %v", *snap.Uuid, *snap.RepositoryUuid)
+			}
+
+			if snap.RepositoryPath == nil {
+				return repositories, customRepositories, fmt.Errorf("No repository path is associated with this snapshot")
+			}
+			composerRepo := composer.Repository{
+				Baseurl: common.ToPtr(h.server.csReposURL.JoinPath(*snap.RepositoryPath).String()),
+				Rhsm:    common.ToPtr(false),
+			}
+
+			if repo.GpgKey != nil && *repo.GpgKey != "" {
+				composerRepo.Gpgkey = repo.GpgKey
+			}
+			if composerRepo.Gpgkey != nil && *composerRepo.Gpgkey != "" {
+				composerRepo.CheckGpg = common.ToPtr(true)
+			}
+			composerRepo.ModuleHotfixes = repo.ModuleHotfixes
+			composerRepo.CheckRepoGpg = repo.MetadataVerification
+			repositories = append(repositories, composerRepo)
+
+			// Don't enable custom repositories, as they require further setup to be useable.
+			customRepo := composer.CustomRepository{
+				Id:      *snap.RepositoryUuid,
+				Name:    repo.Name,
+				Baseurl: &[]string{*snap.Url},
+				Enabled: common.ToPtr(false),
+			}
+			if repo.GpgKey != nil && *repo.GpgKey != "" {
+				customRepo.Gpgkey = &[]string{*repo.GpgKey}
+				customRepo.CheckGpg = common.ToPtr(true)
+			}
+			customRepo.ModuleHotfixes = repo.ModuleHotfixes
+			customRepo.CheckRepoGpg = repo.MetadataVerification
+			customRepositories = append(customRepositories, customRepo)
+		}
+	}
+
+	ctx.Logger().Debugf("Resolved repository snapshots from template: %v", repositories)
```

Comment:
Just a sidenote, we have just integrated our https://github.com/osbuild/logging library and we are in the transition from `logrus` to `slog`. You can either use `echo` logging API, which is this line so all is good, or use `slog` to do all logging. Just avoid using `logrus` completely, just in case you add some more logging into this patch, thanks.